textattack/search_methods/PSO_algorithm.py

```diff
@@ -0,0 +1,266 @@
+"""
+Reimplementation of search method from Word-level Textual Adversarial Attacking as Combinatorial Optimization
+by Zang et. al
+`<https://www.aclweb.org/anthology/2020.acl-main.540.pdf>`_
+`<https://github.com/thunlp/SememePSO-Attack>`_
+"""
+
+from copy import deepcopy
+
+import numpy as np
+
+from textattack.search_methods import SearchMethod
+
+
+class PSOAlgorithm(SearchMethod):
+    """
+    Attacks a model with word substiutitions using a PSO algorithm.
+
+    Args:
+        pop_size (:obj:`int`, optional): The population size. Defauls to 20.
+        max_iters (:obj:`int`, optional): The maximum number of iterations to use. Defaults to 50.
+    """
+
+    def __init__(
+        self, pop_size=60, max_iters=20,
+    ):
+        self.max_iters = max_iters
+        self.pop_size = pop_size
+        self.search_over = False
+
+    def generate_population(self, x_orig, neighbors_list, neighbors_len):
+        h_score, w_list = self.gen_h_score(x_orig, neighbors_len, neighbors_list)
+        return [self.mutate(x_orig, h_score, w_list) for _ in range(self.pop_size)]
+
+    def mutate(self, x_cur, w_select_probs, w_list):
+        rand_idx = np.random.choice(len(w_select_probs), 1, p=w_select_probs)[0]
+        return x_cur.replace_word_at_index(rand_idx, w_list[rand_idx])
+
+    def gen_h_score(self, x, neighbors_len, neighbors_list):
+
+        w_list = []
+        prob_list = []
+        for i, orig_w in enumerate(x.words):
+            if neighbors_len[i] == 0:
+                w_list.append(orig_w)
+                prob_list.append(0)
+                continue
+            p, w = self.gen_most_change(x, i, neighbors_list[i])
+            w_list.append(w)
+            prob_list.append(p)
+
+        prob_list = self.norm(prob_list)
+
+        h_score = prob_list
+        h_score = np.array(h_score)
+        return h_score, w_list
+
+    def norm(self, n):
+
+        tn = []
+        for i in n:
+            if i <= 0:
+                tn.append(0)
+            else:
+                tn.append(i)
+        s = np.sum(tn)
+        if s == 0:
+            for i in range(len(tn)):
+                tn[i] = 1
+            return [t / len(tn) for t in tn]
+        new_n = [t / s for t in tn]
+
+        return new_n
+
+    # for un-targeted attacking
```

Comment:
is this just for un-targeted attacking? won't "goal results" work for any goal?

Reply:
it can be used for targeted attacking, but we should choose either the untargeted attacking or targeted one for this recipe, and isn't the untargeted attack the default one for demo?